MouseBot and Direction Order Change

joshuasrjc opened the pull request:
# Changed
### enum direction

The order of the direction enum was changed from:  
{ NW , N , NE , W , NONE , E , SW , S , SE }  
to  
{ N, NE, E, SE, S, SW, W, NW, NONE }  
In addition, north is now the positive Y direction
### getNeighborNode

The getNeightborNode function in Maze depended on the old direction order, and was changed to use the offset overload function described below.
# Added
### MouseBot Class

Added a mouse class that keeps track of position and direction. It also has functions to move around.
### Direction overloads

Two overload functions using the direction enum were added.  
Rotate: dir1 % dir2 now returns dir1, rotated in the direction of dir2, with north being straight ahead.  
Offset: pos + dir now returns a position, offset in the direction of dir by 1 unit in each direction.
# Tested
## MouseBot

The movement functions of MouseBot have been tested, and works as expected.
## The maze class has not been tested.

joshuasrjc committed:
Created MouseBot

Created the MouseBot Class and added some functions. Not fully
implemented. Not tested.

joshuasrjc committed:
Direction Order Change

Changed the order of the directions to start at north = 0, and rotate
clockwise. Added an overload function so that Pos+ direction = a
position offset in that direction by 1 unit. Changed the getNeighborNode
in Maze to use this new method of offsetting.

joshuasrjc committed:
MouseBot Tested

The MouseBot class has been tested, and basic functions have been
implemented. North (or up) is now the positive Y direction, the idea
being that the robot could assume it starts facing north, and then moves
into positive Y.  The Maze class and pathfinding algorithm have not been
tested with the new direction enum.

joshuasrjc committed:
Cleanup

Cleaned up some changes I made while testing / debugging.

blhough commented on Vector.cpp:
```diff
@@ -1,8 +1,21 @@
 
 namespace Micromouse
 {
+	direction operator%(const direction &dir1, const direction &dir2)
```
I don't think this works as you intended
doing it by hand I get
`NE % S = SW`
`W % W = S`
`NW % N = NW`

joshuasrjc replied:
That works exactly as I intended. It's really easy for me to visualize, but difficult to write out. Suppose I'm facing direction X and I want to turn 45 degrees to the right. I would use X = X % NE, since NE is 45 degrees to the right of north. If I wanted to turn around, I would use X = X % S since south is in the opposite direction of north.

RosyMapleMoth replied:
Ok it took a bit of scratch paper, but I totally get how this works and I think i can write a relatively simple doc explaining it. EDIT Now that I reread what joshua said above. I see it explains it pretty well.

blhough replied:
It has been suggested to change it to an overload of dir + dir so it is more obvious what it does. Also then dir - dir could be added. What do you think @RosyMapleMoth

RosyMapleMoth replied:
I really think it makes more sense from a clarity point of view to make it a new function rather than an overload. but I do think it would make more sense as a + overload then a % overload. as far as adding a dir - dir we really don't need to for any reason, and the more I think about it the more of a headache it sounds like to have different people using dir - dir and dir + dir in different parts of the code.

TL;DR I think it should be it's own function or dir + dir, and we shouldn't have a dir - dir

RosyMapleMoth commented on Vector.cpp:
```diff
@@ -63,7 +91,7 @@ namespace Micromouse
 
-			// no Node should exist at two odd coordinates
+			// no Pos should have a position at two odd coordinates
 			assert( ( _x % 2 != 1 ) && ( _y % 2 != 1 ) );
```
This was already an error with this code but this will fail whenever Pos has an odd coordinate, not when both of its coordinates are odd.

(x % 2 != 1) = false | x = any odd number 
(y % 2 != 1) = false | y = any odd number 

false && true = false
true && false = false

only 
true && true = true

blhough replied:
good catch!
@joshuasrjc will you change it to
`assert( ( _x % 2 == 0 ) || ( _y % 2 == 0 ) );`
please

RosyMapleMoth replied:
Yeah Sure. EDIT : Oops I totally miss read that comment.

joshuasrjc replied:
Fixed!

joshuasrjc committed:
Direction Operators Changed

Changed the % operator to the + operator for rotating directions.
Changed the functionality of the ++ operator for directions to include
the NONE direction. Fixed a for loop in Maze that was dependent on the
old direction order. Fixed validateSelf() in Pos to work as intended.

joshuasrjc committed:
getFacing() Function

Added the getFacing() function to MouseBot.

RosyMapleMoth commented:
Looks good Merging !
